driver/tab_completion.py

```diff
@@ -0,0 +1,120 @@
+
+
+def complete_planner_args(prefix, parsed_args, **kwargs):
+    util.set_default_build(parsed_args)
+
+    # Get some information from planner_args before it is deleted in split_planner_args().
+    planner_args = parsed_args.planner_args
+    num_planner_args = len(planner_args)
+    double_dash_in_options = "--" in planner_args
+
+    current_mode = util.split_planner_args(parsed_args)
+    num_filenames = len(parsed_args.filenames)
+    has_only_filename_options = (num_filenames == num_planner_args)
+    has_only_filename_or_double_dash_options = (num_filenames + int(double_dash_in_options) == num_planner_args)
+    can_use_double_dash = (1 <= num_planner_args <= 2) and has_only_filename_or_double_dash_options
+
+    completions = {}
+
+    if can_use_double_dash:
+        completions["--"] = ""
+
+    if parsed_args.filenames or double_dash_in_options:
+        if current_mode == "search":
+            completions["--translate-options"] = ""
+
+            downward = get_search_executable(parsed_args.build, exit_on_failure=False)
+            if downward and downward.exists():
+                completions.update(_get_completions_from_downward(
+                    downward, parsed_args.search_options, prefix))
+        else:
+            completions["--search-options"] = ""
+
+            translator = get_translate_executable(parsed_args.build, exit_on_failure=False)
+            if translator and translator.exists():
+                completions.update(_get_completions_from_translator(
+                    translator, parsed_args.translate_options, prefix))
+
+    if has_only_filename_options and len(parsed_args.filenames) < 2:
+        file_completer = argcomplete.FilesCompleter()
+        completions.update({f: "" for f in file_completer(prefix, **kwargs)})
+
+    return completions
+
+
+def _get_field_separators(env):
+    entry_separator = env.get("IFS", "\n")
+    help_separator = env.get("_ARGCOMPLETE_DFS")
+    if env.get("_ARGCOMPLETE_SHELL") == "zsh":
+        # Argcomplete always uses ":" on zsh, even if another value is set in
+        # _ARGCOMPLETE_DFS.
+        help_separator = ":"
+    return entry_separator, help_separator
+
+
+def _split_argcomplete_ouput(content, entry_separator, help_separator):
+    suggestions = {}
+    for line in content.split(entry_separator):
+        if help_separator and help_separator in line:
+            suggestion, help = line.split(help_separator, 1)
+            suggestions[suggestion] = help
+        else:
+            suggestions[line] = ""
+    return suggestions
+
+
+def _call_argcomplete(python_file, comp_line, comp_point):
+    with tempfile.NamedTemporaryFile(mode="r") as f:
+        env = os.environ.copy()
+        env["COMP_LINE"] = comp_line
+        env["COMP_POINT"] = str(comp_point)
+        env["_ARGCOMPLETE"] = "1"
+        env["_ARGCOMPLETE_STDOUT_FILENAME"] = f.name
+        subprocess.check_call([sys.executable, python_file], env=env)
```

Comment:
The only other place where the translator is called that I could find was in `run_components.py` where it is also called with `sys.executable` (https://github.com/aibasel/downward/blob/main/driver/run_components.py#L65). We cannot simply reuse this code, because it does a lot of additional stuff: setting up resource limits, handling how output is printed, etc. It also takes an `args` argument that is assumed to have arguments for the call in place. Internally, it uses `get_error_output_and_returncode` which is meant for user-facing execution (e.g., it prints the call string to the log which would be a problem for us here because it would be interpreted as a completion suggestion). I don't think the overlap is big enough that we have to extract some common functionality here.

As to the point about using the shebang: I thought the shebang is only considered if you use subprocess with `shell=True` which I thought to be a security red flag. I see the point of overwriting the shebang this way, but I don't think this is necessarily a problem. Say I have a Python venv and call the driver with the Python version from that venv (without activating it). Then the translator keeps using the venv, which I wouldn't find surprising as a user who doesn't know about the internals of Fast Downward. I think what I'm trying to argue is that if `sys.executable` and the executable on the PATH are not the same, then the user already specified a preference for one over the other in the call to the driver so it is OK to keep using it.

Reply:
Thanks! I looked one level less deeply into the code; not into the code that calls the translator but the code that finds the translator. Because that function was called "get_executable", I assumed that this is then used as an executable, but actually it isn't any more. I see now that this was changed as part of issue733 in order to be able to do experiments that compare the translator on different Python versions by calling the driver with different Python versions.

No, the shebang and shell=True are not related. (And indeed shell=True needs a lot of care to avoid security issues.) Linux has no notion of "file extensions" inside, a filename is just a string. So if I run a file as an executable, Linux needs some way to figure out how to run it. So it looks inside the file. If the file has a recognized binary format (usually ELF), then the kernel knows how to run ELF binaries. Otherwise it looks at the first line to find out which other executable to use in order to invoke the program. It's an OS mechanism, not a shell mechanism. (This is different from Windows, which registers specific file extensions with specific programs and has no real notion of an executable script.)

I'm OK with continuing to use get_executable, though I do find it surprising behaviour. (I understand the use case from issue733, although it could also have been achieved by setting PATH.) I don't know other programs that behave in this way when calling another Python program that is intended as a separate executable. Because for me it's a very unintuitive thing to do, I don't like encoding this logic twice.

I think as a user I would be surprised if a venv takes precedence over the installed version if I don't activate it. Perhaps this depends on whether you think of the driver and translator as independent programs.

Reply:
(BTW, looks like Linux now also has a registration mechanism that *can* use file extensions as part of the `binfmt_misc` infrastructure. Not relevant to what we're doing here, but it means that "Linux has no notion of file extensions" is not true; binfmt_misc does support dispatch based on file extensions, although I haven't seen an example of this in use.)

Reply:
In any case, I see why you find this difficult to untangle. In my view, part of the reason for this is that the {_,}get_..._executable functions don't have a very good design with the way they handle errors in two different ways depending on an argument passed in. I think in a cleaner design they would not have this argument and should instead raise an exception in cases where they fail. These exceptions could then be handled in different ways by the different callers, although and the `if downward` etc. tests avoided. Then if we want to call the translator as `[sys.executable, translator_python_file]`, I'd have this list be what `get_translator_executable` (perhaps with a different name; it's not a good name currently as described, but also not a good name if it returns this list) returns. And rather than a method for calling generic Python code, I wouldn't parameterize it with the translator source file, but always keep the information on how to call the translator as a unit.

Reply:
We tried to write another version that uses `get_*_command()` instead of `get_*_executable()` and trows exceptions there. This now has the bundling of `[sys.executable, translate.py]` at only one place, but now we repeat some error handling code instead. We also moved the lookup of the executable further inside to `_get_completions_from_translator`. We didn't move it here (`_call_argcomplete`), because the surrounding function `_get_completions_from_translator`also needs it to compute the line for which we want to get completions.

Reply:
The commit that addresses these comments is https://github.com/aibasel/downward/pull/226/commits/a239859f653d779ca6b9b1b534c4217c5dd094c2. Does it resolve this comment?